rust/src/ikev1/parser.rs

```diff
@@ -0,0 +1,758 @@
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use nom::*;
+use core::fmt;
+use crate::log::*;
+use nom::number::streaming::{be_u16, be_u8, be_u32, be_u64};
+use std::collections::HashSet;
+use crate::common::to_hex;
+
+// Generic ISAKMP "Container" structs
+#[repr(u8)]
+#[derive(Copy, Clone)]
+pub enum ExchangeType {
+    None = 0,
+    Base = 1,
+    IdentityProtection = 2,
+    AuthenticationOnly = 3,
+    Aggressive = 4,
+    Informational = 5,
+    Transaction = 6,
+    QuickMode = 32,
+    NewGroupMode = 33
+}
+
+impl fmt::Display for ExchangeType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ExchangeType::Base => write!(f, "Base"),
+            ExchangeType::IdentityProtection => write!(f, "Identity Protection"),
+            ExchangeType::AuthenticationOnly => write!(f, "Authentication Only"),
+            ExchangeType::Aggressive => write!(f, "Aggressive"),
+            ExchangeType::Informational => write!(f, "Informational"),
+            ExchangeType::Transaction => write!(f, "Transaction (Config Mode)"),
+            ExchangeType::QuickMode => write!(f, "Quick Mode"),
+            ExchangeType::NewGroupMode => write!(f, "New Group Mode"),
+            _ => write!(f, "Unknown Exchange Type")
+        }
+    }
+}
+
+pub struct IsakmpHeader {
+    pub init_spi: u64,
+    pub resp_spi: u64,
+    pub next_payload: u8,
+    pub maj_ver: u8,
+    pub min_ver: u8,
+    pub exch_type: u8,
+    pub flags: u8,
+    pub msg_id: u32,
+    pub length: u32,
+}
+
+pub struct IsakmpPayloadHeader {
+    pub next_payload: u8,
+    pub reserved: u8,
+    pub payload_length: u16,
+}
+
+pub struct IsakmpPayload<'a> {
+    pub payload_header: IsakmpPayloadHeader,
+    pub data: &'a [u8],
+}
+
+// IKEV1 specific payloads
+
+// 1 -> Security Association
+pub struct SecurityAssociationPayload<'a> {
+    pub domain_of_interpretation: u32,
+    pub situation: Option<&'a [u8]>,
+    pub data: Option<&'a [u8]>
+}
+
+// 2 -> Proposal
+pub struct ProposalPayload<'a> {
+    pub proposal_number: u8,
+    pub proposal_type: u8,
+    pub spi_size: u8,
+    pub number_transforms: u8,
+    pub spi: &'a [u8],
+    pub data: &'a [u8]
+}
+
+// 3 -> Transform
+pub struct TransformPayload<'a> {
+    pub transform_number: u8,
+    pub transform_type: u8,
+    pub sa_attributes: &'a [u8]
+}
+
+// 4 -> Key Exchange
+pub struct KeyExchangePayload<'a> {
+    pub key_exchange_data: &'a [u8],
+}
+
+// 5 -> Identification
```

Comment:
Are these TODOs ?

Reply:
not sure :)

didn't need the payload-types now but wanted to have them at least noted. Perhaps there is in the future the need to add them.